tarofreighter/chain_porter.go

```diff
@@ -502,6 +559,108 @@ func (p *ChainPorter) transferReceiverProof(pkg *sendPackage) error {
 	return nil
 }
 
+// timeSinceLastDeliveryAttempt calculates time duration which has elapsed since
+// the last delivery attempt.
+func timeSinceLastDeliveryAttempt(timestamps []time.Time) time.Duration {
+	// If there are no previous parcel delivery attempts, then we'll
+	// return early.
+	if len(timestamps) == 0 {
+		return time.Duration(0)
+	}
+
+	// Otherwise we'll select the latest timestamp and compute the surpassed
+	// time relative to the current time.
+
+	// Get the latest timestamp without assuming order.
+	latestTimestamp := timestamps[0]
+	for _, timestamp := range timestamps {
+		if timestamp.After(latestTimestamp) {
+			latestTimestamp = timestamp
+		}
+	}
+
+	return time.Since(latestTimestamp)
+}
+
+// backoffTransferReceiverProof attempts to transfer the receiver proof to the
+// receiver using a backoff strategy. The backoff strategy is used to ensure
+// that we don't spam the receiver with proof delivery attempts.
+func (p *ChainPorter) backoffTransferReceiverProof(ctx context.Context,
+	pkg *sendPackage, receiverProof *proof.AnnotatedProof) error {
+
+	var (
+		backoff          = p.cfg.ReceiverProofTransferConfig.InitialBackoff
+		numTries         = p.cfg.ReceiverProofTransferConfig.NumTries
+		maxBackoff       = p.cfg.ReceiverProofTransferConfig.MaxBackoff
+		err        error = nil
+	)
+
+	for i := 0; i < numTries; i++ {
+		// Don't delay at first try. Wait before reattempting
+		// delivery if backoff specified.
+		if i > 0 && backoff != 0 {
+			err = p.wait(backoff)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Before attempting to deliver the proof, log that
+		// an attempted delivery is about to occur.
+		err = p.cfg.ExportLog.StoreParcelDeliveryAttempt(
+			ctx, pkg.OutboundPkg.NewAnchorPoint,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to log proof delivery "+
+				"attempt: %w", err)
+		}
+
+		err = p.cfg.ProofCourier.DeliverProof(
```

Comment:
We also only want to call this once really, as this'll handle the set up for hash mail creation, etc.

Reply:
I see what you mean, that function call also initialises the mailboxes. I intend on refactoring this function and keeping the mailbox states on the `ProofCourier`. Hopefully I can setup aperture before that.